File: BSharp/ClientApp/src/app/data/dto/role.ts
import { DtoForSaveKeyBase } from './dto-for-save-key-base';
import { Permission, PermissionForSave } from './permission';
import { RoleMembershipForSave, RoleMembership } from './role-membership';
import { RequiredSignatureForSave, RequiredSignature } from './required-signature';
import { SettingsForClient } from './settings';
import { DtoDescriptor } from './metadata';
import { TenantWorkspace } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { DtoKeyBase } from './dto-key-base';

export class RoleForSave<TPermission = PermissionForSave, TRequiredSignature = RequiredSignatureForSave,
    TRoleMembership = RoleMembershipForSave> extends DtoForSaveKeyBase {
    Name: string;
    Name2: string;
    Code: string;
    IsPublic: boolean;
    Permissions: TPermission[];
    Signatures: TRequiredSignature[];
    Members: TRoleMembership[];
}

export class Role extends RoleForSave<Permission, RequiredSignature, RoleMembership> {
    IsActive: boolean;
    CreatedAt: string;
    CreatedById: number | string;
    ModifiedAt: string;
    ModifiedById: number | string;
}

const _select = ['', '2', '3'].map(pf => 'Name' + pf);
let _currentLang: string;
let _settings: SettingsForClient;
let _cache: DtoDescriptor;

export function metadata_Role(ws: TenantWorkspace, trx: TranslateService, _subtype: string): DtoDescriptor {
  // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
  if (trx.currentLang !== _currentLang && ws.settings !== _settings) {
    _currentLang = trx.currentLang;
    _settings = ws.settings;
    _cache = {
      select: _select,
      orderby: ws.isSecondaryLanguage ? `${_select[1]},${_select[0]}` : ws.isTernaryLanguage ? `${_select[2]},${_select[0]}` : _select[0],
      format: (item: DtoKeyBase) => ws.getMultilingualValueImmediate(item, _select[0]),
      properties: {
        Name: { control: 'text', label: trx.instant('Name') + ws.primaryPostfix },
        Name2: { control: 'text', label: trx.instant('Name') + ws.secondaryPostfix },
        Name3: { control: 'text', label: trx.instant('Name') + ws.ternaryPostfix },
        Code: { control: 'text', label: trx.instant('Code') },
        IsPublic: { control: 'boolean', label: trx.instant('Role_IsPublic') },
        IsActive: { control: 'boolean', label: trx.instant('IsActive') },
        CreatedAt: { control: 'datetime', label: trx.instant('CreatedAt') },
        CreatedBy: { control: 'navigation', label: trx.instant('CreatedBy'), type: 'LocalUser', foreignKeyName: 'CreatedById' },
        ModifiedAt: { control: 'datetime', label: trx.instant('ModifiedAt') },
        ModifiedBy: { control: 'navigation', label: trx.instant('ModifiedBy'), type: 'LocalUser', foreignKeyName: 'ModifiedById' }
      }
    };
  }

  return _cache;
}

File: BSharp/Controllers/DTO/Role.cs
using BSharp.Controllers.Misc;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace BSharp.Controllers.DTO
{
    public class RoleForSave<TPermission, TRequiredSignature, TRoleMembership> : DtoForSaveKeyBase<int?>
    {
        [BasicField]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [MultilingualDisplay(Name = "Name", Language = Language.Primary)]
        public string Name { get; set; }

        [BasicField]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [MultilingualDisplay(Name = "Name", Language = Language.Secondary)]
        public string Name2 { get; set; }

        [BasicField]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [Display(Name = "Code")]
        public string Code { get; set; }

        [BasicField]
        [Display(Name = "Role_IsPublic")]
        public bool IsPublic { get; set; }

        [NavigationProperty(ForeignKey = nameof(PermissionForQuery.RoleId))]
        [Display(Name = "Permissions")]
        public List<TPermission> Permissions { get; set; } = new List<TPermission>();

        [NavigationProperty(ForeignKey = nameof(RequiredSignatureForQuery.RoleId))]
        [Display(Name = "Signatures")]
        public List<TRequiredSignature> Signatures { get; set; } = new List<TRequiredSignature>();

        [NavigationProperty(ForeignKey = nameof(RoleMembershipForQuery.RoleId))]
        [Display(Name = "Members")]
        public List<TRoleMembership> Members { get; set; } = new List<TRoleMembership>();
    }

    public class RoleForSave : RoleForSave<PermissionForSave, RequiredSignatureForSave, RoleMembershipForSave>
    {

    }

    public class Role<TPermission, TRequiredSignature, TRoleMembership> : RoleForSave<TPermission, TRequiredSignature, TRoleMembership>
    {
        [BasicField]
        [Display(Name = "IsActive")]
        public bool? IsActive { get; set; }

        [Display(Name = "CreatedAt")]
        public DateTimeOffset? CreatedAt { get; set; }

        [ForeignKey]
        [Display(Name = "CreatedBy")]
        public int? CreatedById { get; set; }

        [Display(Name = "ModifiedAt")]
        public DateTimeOffset? ModifiedAt { get; set; }

        [ForeignKey]
        [Display(Name = "ModifiedBy")]
        public int? ModifiedById { get; set; }
    }

    public class Role : Role<Permission, RequiredSignature, RoleMembership>
    {

    }

    public class RoleForQuery : Role<PermissionForQuery, RequiredSignatureForQuery, RoleMembershipForQuery>, IAuditedDto
    {
        [NavigationProperty(ForeignKey = nameof(CreatedById))]
        public LocalUserForQuery CreatedBy { get; set; }

        [NavigationProperty(ForeignKey = nameof(ModifiedById))]
        public LocalUserForQuery ModifiedBy { get; set; }
    }
}
